core/services/pipeline/orm.go

```diff
@@ -0,0 +1,360 @@
+package pipeline
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
+	"gopkg.in/guregu/null.v4"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+//go:generate mockery --name ORM --output ./mocks/ --case=underscore
+
+type ORM interface {
+	CreateSpec(taskDAG TaskDAG) (int32, error)
+	CreateRun(jobID int32, meta map[string]interface{}) (int64, error)
+	ProcessNextUnclaimedTaskRun(f func(jobID int32, taskRun TaskRun, predecessors []TaskRun) Result) (bool, error)
+	ListenForNewRuns() (*utils.PostgresEventListener, error)
+	AwaitRun(ctx context.Context, runID int64) error
+	ResultsForRun(runID int64) ([]Result, error)
+	// Runs(jobID int32, offset, limit int) ([]Run, error)
+
+	FindBridge(name models.TaskType) (models.BridgeType, error)
+}
+
+type orm struct {
+	db  *gorm.DB
+	uri string
+}
+
+var _ ORM = (*orm)(nil)
+
+func NewORM(db *gorm.DB, uri string) *orm {
+	return &orm{db, uri}
+}
+
+func (o *orm) CreateSpec(taskDAG TaskDAG) (int32, error) {
+	var specID int32
+	err := utils.GormTransaction(o.db, func(tx *gorm.DB) (err error) {
+		now := time.Now()
+
+		// Create the pipeline spec
+		spec := Spec{
+			DotDagSource: taskDAG.DOTSource,
+			CreatedAt:    now,
+		}
+		err = tx.Create(&spec).Error
+		if err != nil {
+			return err
+		}
+		specID = spec.ID
```

Comment:
Do we need this? Doesnt seem to save much

Reply:
Debris, will clean